invenio/modules/relationships/engines/sqlalchemy.py

```diff
@@ -0,0 +1,117 @@
+from uuid import uuid4
+
+from sqlalchemy_utils import UUIDType
+
+from invenio.ext.sqlalchemy import db, utils
+
+
+class SQLAlchemyNode(object):
+
+    """TODO"""
+
+    def _edges(self, inwards=True, outwards=True, link_type=None):
+        """TODO"""
+        query = SQLAlchemyEdge.query
+
+        if inwards and outwards:
+            query = query.filter((SQLAlchemyEdge.id_to == self['_id']) |
+                                 (SQLAlchemyEdge.id_from == self['_id']))
+        elif inwards:
+            query = query.filter(SQLAlchemyEdge.id_to == self['_id'])
+        elif outwards:
+            query = query.filter(SQLAlchemyEdge.id_from == self['_id'])
+
+        if link_type:
+            query = query.filter(SQLALchemyEdge.link_type == link_type)
+
+        return query
+
+    def in_edges(self, link_type=None):
+        """TODO"""
+        edges = self._edges(outwards=False, link_type=link_type)
+
+        return edges.all()
+
+    def in_degree(self, link_type=None):
+        """TODO"""
+        return len(self.in_edges())
+
+    def out_edges(self, link_type=None):
+        """TODO"""
+        edges = self._edges(inwards=False, link_type=link_type)
+
+        return edges.all()
+
+    def out_degree(self, link_type=None):
+        """TODO"""
+        return len(self.out_edges())
+
+    def predecessors(self, link_type=None):
+        """TODO"""
+        in_edges = self.in_edges(link_type=link_type)
+
+        return [edge.from_node() for edge in in_edges]
+
+    def successors(self, link_type=None):
+        """TODO"""
+        out_edges = self.out_edges(link_type=link_type)
+
+        return [edge.to_node() for edge in out_edges]
+
+
+class SQLAlchemyEdge(db.Model):
+
+    """Represent a graph edge."""
+
+    __tablename__ = "relationship"
+
+    uuid = db.Column(UUIDType(binary=False), primary_key=True)
+
+    model_from = db.Column(db.String(255), nullable=True, index=True)
+    id_from = db.Column(db.Integer(unsigned=True), nullable=True, index=True)
+
+    link_type = db.Column(db.String(255), index=True)
+    attributes = db.Column(db.JSON)
+
+    model_to = db.Column(db.String(255), nullable=True, index=True)
+    id_to = db.Column(db.Integer(unsigned=True), nullable=True, index=True)
+
```

Comment:
No, no issues with that. Will investigate.

I've read [this](http://alextechrants.blogspot.ch/2013/11/10-common-stumbling-blocks-for.html) on this topic, point 2. I read that as "usually they aren't needed, but newbies defined them anyway". I thought that in this case is OK since there's a little logic in them.